tests/component/test_stream_readers.py

```diff
@@ -545,3 +547,591 @@ def test___power_binary_reader___read_many_sample_with_wrong_dtype___raises_erro
+    # the second 8-bit port. Instead of the data being 0x00F0, it would be 0xF000 (little-endian).
+    # Anyways, its harder to reason about, so we'll only use 8 lines. DAQ will happily extend the
+    # data to the larger type.
+    task.di_channels.add_di_chan(
+        flatten_channel_string(sim_6363_device.di_lines[32:40].name),
+        line_grouping=LineGrouping.CHAN_PER_LINE,
+    )
+    return task
+
+
+@pytest.fixture
+def di_single_channel_port_uint32_task(
+    task: nidaqmx.Task, sim_6363_device: nidaqmx.system.Device
+) -> nidaqmx.Task:
+    # 6363 port 0 has 32 lines
+    task.di_channels.add_di_chan(
+        sim_6363_device.di_ports[0].name,
+        line_grouping=LineGrouping.CHAN_FOR_ALL_LINES,
+    )
+    return task
+
+
+@pytest.fixture
+def di_multi_channel_port_uint32_task(
+    task: nidaqmx.Task, sim_6363_device: nidaqmx.system.Device
+) -> nidaqmx.Task:
+    # 6363 port 0 has 32 lines
+    task.di_channels.add_di_chan(
+        flatten_channel_string(sim_6363_device.di_lines[:32].name),
+        line_grouping=LineGrouping.CHAN_PER_LINE,
+    )
+    return task
+
+
+def _get_expected_digital_data_for_sample(num_lines: int, sample_number: int) -> int:
+    # Simulated digital signals "count" from 0 in binary within each group of 8 lines.
+    result = sample_number
+    for _ in range(num_lines // 8):
+        result = (result << 8) | sample_number
+
+    line_mask = (2**num_lines) - 1
+    return result & line_mask
+
+
+def _get_expected_digital_data(num_lines: int, num_samples: int) -> List[int]:
+    return [
+        _get_expected_digital_data_for_sample(num_lines, sample_number)
+        for sample_number in range(num_samples)
+    ]
+
+
+def _get_expected_digital_port_data_for_sample(num_lines: int, sample_number: int) -> List[int]:
+    expected_data = _get_expected_digital_data_for_sample(num_lines, sample_number)
+    # port data is strange, you get full port data masked for only the line in the channel
+    return [expected_data & (1 << line) for line in range(num_lines)]
+
+
+def _get_expected_digital_port_data(num_lines: int, num_samples: int) -> List[List[int]]:
+    result = []
+    # again, port data is strange. it is line-major with the same odd masking/port format as above
+    for line in range(num_lines):
+        result.append(
+            [datum & (1 << line) for datum in _get_expected_digital_data(num_lines, num_samples)]
+        )
+    return result
+
+
+def _bool_array_to_int(bool_array: numpy.ndarray) -> int:
+    result = 0
+    # Simulated data is little-endian
+    for bit in reversed(bool_array):
+        result = (result << 1) | bit
+    return result
+
+
+def test___digital_single_channel_reader___read_one_sample_one_line___returns_valid_samples(
+    di_single_line_task: nidaqmx.Task,
+) -> None:
+    reader = nidaqmx.stream_readers.DigitalSingleChannelReader(di_single_line_task.in_stream)
+    num_lines = di_single_line_task.channels.di_num_lines
```

Comment:
This looks like it's getting the total number of lines, but `task.channels.di_num_lines` returns the number of lines per channel and errors if they're not the same. It happens to work because there is only one channel.

I see the multi-channel multi-line test cases are setting `num_lines = task.number_of_channels`, which makes a different assumption: that each channel contains one line.

I think this is kind of misleading, but more importantly it highlights the lack of coverage of less trivial multi-line multi-channel use cases. DAQmx lets you group an arbitrary number of lines into each channel. It's easy to go overboard here, but perhaps we could at least test the most obvious case, chan0="Dev1/port0" (32) chan1="Dev1/port1" (8) chan2="Dev1/port2" (8). 

It's ok to split this into a separate PR.

Reply:
This test case naturally appeared out of my refactor to move to ports instead of lines for multi-channel tests. `di_multi_channel_port_uint32_task` is your exact example.